brainscore_language/models/huggingface.py

```diff
@@ -99,59 +103,97 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
+                representations = self.output_to_representations(layer_representations, stimuli_coords=stimuli_coords)
                 output['neural'].append(representations)
 
+        # merge over text parts
         output['behavior'] = merge_data_arrays(output['behavior']).sortby('part_number') if output['behavior'] else None
         output['neural'] = merge_data_arrays(output['neural']).sortby('part_number') if output['neural'] else None
         return output
 
-    def estimate_reading_times(self, base_output):
+    def _setup_hooks(self):
+        """ set up the hooks for recording internal neural activity from the model (aka layer activations) """
+        hooks = []
+        layer_representations = OrderedDict()
+        for (recording_target, recording_type) in self.neural_recordings:
+            layer_name = self.region_layer_mapping[recording_target]
+            layer = self._get_layer(layer_name)
+            hook = self._register_hook(layer, name=(recording_target, layer_name),
+                                       target_dict=layer_representations)
+            hooks.append(hook)
+        return hooks, layer_representations
+
+    def output_to_representations(self, layer_representations, stimuli_coords):
+        representation_values = np.concatenate([
+            # Choose to use last token (-1) of values[batch, token, unit] to represent passage.
+            values[:, -1:, :].squeeze(0) for values in layer_representations.values()],
+            axis=-1)  # concatenate along neuron axis
+        neuroid_coords = {
+            'layer': ('neuroid', np.concatenate([[layer] * values.shape[-1]
+                                                 for (recording_target, layer), values in
+                                                 layer_representations.items()])),
+            'region': ('neuroid', np.concatenate([[recording_target] * values.shape[-1]
+                                                  for (recording_target, layer), values in
+                                                  layer_representations.items()])),
+            'neuron_number_in_layer': ('neuroid', np.concatenate(
+                [np.arange(values.shape[-1]) for values in layer_representations.values()])),
+        }
+        neuroid_coords['neuroid_id'] = 'neuroid', functools.reduce(defchararray.add, [
+            neuroid_coords['layer'][1], '--', neuroid_coords['neuron_number_in_layer'][1].astype(str)])
+        representations = NeuroidAssembly(
+            representation_values,
+            coords={**stimuli_coords, **neuroid_coords},
+            dims=['presentation', 'neuroid'])
+        return representations
+
+    def estimate_reading_times(self, base_output: CausalLMOutput):
         """
         :param base_output: the neural network's output
-        :return: perplexity (calculated as cross entropy) as a proxy for reading times
+        :return: cross entropy as a proxy for reading times, following Smith & Levy 2013
```

Comment:
"surprisal (in bits) as a proxy for reading times"
(ambiguity over log-base can be an issue!)